onnxruntime/core/providers/partitioning_utils.cc

```diff
@@ -3,97 +3,173 @@
+  std::deque<const Node*> nodes_to_process{};
+  // nodes that will be processed when considering the next partition node group
+  std::deque<const Node*> nodes_to_process_with_next_group{};
+
+  // initialize in-degrees and find root nodes
+  for (const auto& node : graph_viewer.Nodes()) {
+    const auto node_input_edge_count = node.GetInputEdgesCount();
+    in_degree.insert({node.Index(), node_input_edge_count});
+    if (node_input_edge_count == 0) {
+      nodes_to_process.push_back(&node);
     }
   }
 
-  // prefer unsupported nodes first. this will increase the number of inputs potentially available to the first
-  // partition handled by this EP.
-  bool processing_supported_nodes = false;
+  std::vector<const Node*> supported_group{};
+  // the partition node group's frontier is the aggregate of its nodes' output nodes
+  std::unordered_set<const Node*> supported_group_frontier{};
+
+  auto close_group = [&]() {
+    if (!supported_group.empty()) {
+#ifndef NDEBUG
+      if (debug_output) {
+        LOGS_DEFAULT(VERBOSE) << "New partition node group.\n"
+                              << "Unsupported nodes on group frontier: "
+                              << NodeGroupDebugString(nodes_to_process_with_next_group, true) << "\n"
+                              << "Nodes in group: " << NodeGroupDebugString(supported_group);
+      }
+#endif
 
-  while (!supported_to_visit.empty() || !unsupported_to_visit.empty()) {
-    const Node* current = nullptr;
+      // if no on_group_closed_fn callback was given, keep the partition
+      // otherwise, let the callback determine whether to keep it
+      const bool keep_partition = !on_group_closed_fn || on_group_closed_fn(supported_group);
 
-    // see if we need to flip
-    if ((processing_supported_nodes && supported_to_visit.empty()) ||
-        (!processing_supported_nodes && unsupported_to_visit.empty())) {
-      processing_supported_nodes = !processing_supported_nodes;
+      if (keep_partition) {
+        supported_groups.emplace_back(std::move(supported_group));
+      } else {
+        LOGS_DEFAULT(VERBOSE) << "Discarded partition node group.";
+      }
+
+      supported_group.clear();
+      supported_group_frontier.clear();
+    }
+  };
+
+  while (!nodes_to_process.empty() || !nodes_to_process_with_next_group.empty()) {
+    if (nodes_to_process.empty()) {
+      // we have processed all the nodes that we can while building this partition node group, start a new one
+      close_group();
+      nodes_to_process.swap(nodes_to_process_with_next_group);
       continue;
     }
 
-    // get next node from same partition
-    if (processing_supported_nodes) {
-      current = supported_to_visit.front();
-      supported_to_visit.pop();
-    } else {
-      current = unsupported_to_visit.front();
-      unsupported_to_visit.pop();
+    const Node& node = *nodes_to_process.front();
+    nodes_to_process.pop_front();
+
+    const bool is_node_supported = is_node_supported_fn(node);
+
```

Comment:
Do we still always start with unsupported nodes? It's important to do so as the unsupported nodes may produce outputs that enable supported nodes to form a larger partition.

Reply:
now, we don't necessarily start with unsupported nodes. instead, for a single partition, we process:
- supported nodes - these are included in the current partition
- unsupported nodes that are not downstream of the current partition. the other unsupported nodes can't be processed yet as they form the boundary of the current partition

this also allows the unsupported nodes to be processed and enable other potentially supported nodes to become ready to process.

Reply:
So it's like a wave that flows through the model, starting at the root node/s, that stops when it gets to a point where the only remaining nodes at the edge of the wave are unsupported nodes that are outputs from supported nodes? As the wave goes through it flows over both unsupported and supported nodes, adding the supported nodes to the current partition.

Following that, the next wave starts with the nodes where the previous wave stopped.

Reply:
Seems the logic has changed a lot, maybe we can use a bit more comments here for futrue reader of these code,
For example, check if all the inputs are available is now part of the checking if the node's indegree == 0, ...

Reply:
@skottmckay that's a good way of describing it. I'll add some comments